File: output/gemma4-e4b_IO_1/avatar/Python/Java/atcoder_ABC133_E.java
import java.util.*;
import java.io.*;

public class atcoder_ABC133_E {

    static ArrayList<Integer>[] graph;
    static long MOD = 1_000_000_007L;
    static int K_factor; // Store K globally or pass it

    /**
     * Performs Depth First Search and calculates the product based on the factor K.
     * @param fro The current node.
     * @param to The parent node (to avoid going back immediately).
     * @param f The initial factor/color count for the current node's calculation.
     * @return The calculated product modulo MOD.
     */
    public static long dfs(int fro, int to, int f) {
        long caseNum = f;
        
        // Calculate the factor for the first child.
        // Python logic: colorNum = K - 1 if f == K else K - 2
        // We use K_factor (which holds the input K) for the comparison.
        int colorNum;
        if (f == K_factor) {
            colorNum = K_factor - 1;
        } else {
            colorNum = K_factor - 2;
        }
        
        for (int u : graph[fro]) {
            if (u == to) {
                continue;
            }
            
            // Recursive call: pass the current colorNum as the factor for the child
            long pat = dfs(u, fro, colorNum);
            
            if (pat == 0) {
                caseNum = 0;
                break;
            }
            
            // Update product: caseNum = caseNum * pat % mod
            caseNum = (caseNum * pat) % MOD;
            
            // Decrement the factor for the next child
            colorNum--;
        }
        return caseNum;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        if (!scanner.hasNextInt()) {
            return;
        }
        
        int N = scanner.nextInt();
        int K = scanner.nextInt();
        
        K_factor = K; // Store K for use in DFS
        
        // Initialize graph
        graph = new ArrayList[N];
        for (int i = 0; i < N; i++) {
            graph[i] = new ArrayList<>();
        }
        
        // Read edges
        for (int i = 0; i < N - 1; i++) {
            int a = scanner.nextInt();
            int b = scanner.nextInt();
            
            // Nodes are 1-indexed in input, 0-indexed internally
            graph[a - 1].add(b - 1);
            graph[b - 1].add(a - 1);
        }
        
        scanner.close();
        
        // Initial call: dfs(0, -1, K)
        long ans = dfs(0, -1, K);
        
        System.out.println(ans);
    }
}

File: output/llama3.1-8b_IO_1/avatar/Java/Python/atcoder_ABC133_E.py
from collections import defaultdict, deque

class Node:
    def __init__(self, i):
        self.pattern = 0
        self.index = i
        self.children = []
        self.parent = -1

def solve():
    N = int(input())
    K = int(input())

    nodes = [Node(i) for i in range(N)]

    for _ in range(N-1):
        a, b = map(int, input().split()) - 1
        nodes[a].children.append(b)
        nodes[b].children.append(a)

    def dfs(node, parent, use, nodes):
        node.pattern = K - use
        node.parent = parent
        count = 1
        if parent != -1:
            count += 1
        for child in node.children:
            if child == parent:
                continue
            dfs(nodes[child], node.index, count, nodes)
            count += 1

    def read_ints():
        return map(int, input().split())

    a, b = next(read_ints()) - 1, next(read_ints()) - 1
    nodes[a].children.append(b)
    nodes[b].children.append(a)

    dfs(nodes[0], -1, 0, nodes)

    patterns = 1
    for i in range(N):
        patterns = (patterns * nodes[i].pattern) % 1000000007

    print(patterns)

solve()
